### PR TITLE
added new files for leverj futures version that can handle leverage p…

### PR DESCRIPTION
**Pull request by mitakash**
…arameter

### CODE REVIEW

File: market_maker_keeper/leverjfutures_market_maker_keeper_v2.py [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+        self.precision = -(int(log10(float(quote_increment)))+1)
+
+    def shutdown(self):
+        self.order_book_manager.cancel_all_orders()
+
+    def pair(self):
+        name_to_id_map = {'BTCDAI': '1', 'ETHDAI': '2'}
+        return name_to_id_map[self.arguments.pair.upper()]
+
+    def leverage(self):
```
Comment (MikeHathaway) [inline market_maker_keeper/leverjfutures_market_maker_keeper_v2.py:172]: if leverage is defaulted to 1.0, this entire method can likely be removed

File: market_maker_keeper/leverjfutures_market_maker_keeper_v2.py [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+        self.logger.debug(f'allocation: {allocation}')
+        return allocation
+
+    def our_sell_orders(self, our_orders: list) -> list:
+        return list(filter(lambda order: order.is_sell, our_orders))
+
+    def our_buy_orders(self, our_orders: list) -> list:
+        return list(filter(lambda order: not order.is_sell, our_orders))
+
+    def adjust_target_price(self, target_price):
```
Comment (MikeHathaway) [inline market_maker_keeper/leverjfutures_market_maker_keeper_v2.py:272]: Would be helpful to have typings to differentiate `Wad` or `Price` objects

File: market_maker_keeper/leverjfutures_market_maker_keeper_v2.py [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+from pymaker.numeric import Wad
+from pyexchange.leverjfutures_v2 import LeverjFuturesAPI, Order
+from web3 import Web3, HTTPProvider
+from pymaker.keys import register_keys
+from decimal import *
+
+_context = Context(prec=1000, rounding=ROUND_DOWN)
+
+
+class LeverjMarketMakerKeeper:
```
Comment (MikeHathaway) [inline market_maker_keeper/leverjfutures_market_maker_keeper_v2.py:42]: This class can inherit from: https://github.com/makerdao/market-maker-keeper/blob/master/market_maker_keeper/cex_api.py for further simplification

File: market_maker_keeper/leverjfutures_market_maker_keeper_v2.py [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+            leverage_in_float = 1.0
+        return leverage_in_float
+
+    def token_sell(self) -> str:
+        return self.arguments.pair.upper()[:3]
+
+    def token_buy(self) -> str:
+        return self.arguments.pair.upper()[3:]
+
+    def allocated_balance(self, token: str) -> Wad:
```
Comment (MikeHathaway) [inline market_maker_keeper/leverjfutures_market_maker_keeper_v2.py:184]: you might want to separate the act of accessing a tokens allocation, from it's generation. As used in `synchronize_orders` it looks like you would be regenerating and querying immediately after each other, which would seemingly result in different allocations depending on the time slice

Comment (MikeHathaway) [reply market_maker_keeper/leverjfutures_market_maker_keeper_v2.py:184]: I.e just remove the `.get`, create a variable equal to the return value of `allocated_balance` and call `.get()` on that

File: market_maker_keeper/leverjfutures_market_maker_keeper_v2.py (hunk 2 above)
Comment (MikeHathaway) [reply market_maker_keeper/leverjfutures_market_maker_keeper_v2.py:272]: Also, `assert (isinstance(target_price, Wad))` at the start to catch any argument errors, which simplifys debugging

File: market_maker_keeper/leverjfutures_market_maker_keeper_v2.py [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+        total_available = self.leverj_api.get_plasma_balance(quote_asset_address)
+        # adjust for leverage
+        total_available = Decimal(total_available)*Decimal(self.leverage())
+        self.logger.debug(f'total_available: {total_available}')
+        return self._allocate_to_pair(total_available).get(token)
+
+    def _allocate_to_pair(self, total_available):
+        # total number of instruments across which the total_available balance is distributed
+        # total_available is denominated in quote units
+        total_number_of_instruments = 1
```
Comment (MikeHathaway) [inline market_maker_keeper/leverjfutures_market_maker_keeper_v2.py:196]: Would this ever change?

File: market_maker_keeper/leverjfutures_market_maker_keeper_v2.py [hunk 6]
```diff
@@ -0,0 +1,331 @@
...
+        
+        
+        # buffer_adjustment_factor is a small intentional buffer to avoid allocating the maximum possible. 
+        # the allocated amount is a little smaller than the maximum possible allocation 
+        # and that is determined by the buffer_adjustment_factor
+        buffer_adjustment_factor = Wad.from_number(1.05)
+        
+        base = self.arguments.pair.upper()[:3]
+        quote = self.arguments.pair.upper()[3:]
+        target_price = self.price_feed.get_price()
```
Comment (MikeHathaway) [inline market_maker_keeper/leverjfutures_market_maker_keeper_v2.py:210]: You might want to pass the target price calculated earlier in `synchronize_orders` through as an argument to ensure that all calculations are done using consistent parameters.

File: market_maker_keeper/leverjfutures_market_maker_keeper_v2.py [hunk 7]
```diff
@@ -0,0 +1,331 @@
...
+        product = self.leverj_api.get_product(self.pair())
+        minimum_order_quantity = self.leverj_api.get_minimum_order_quantity(self.pair())
+        minimum_quantity_wad = Wad.from_number(minimum_order_quantity)
+
+        if ((base == product['baseSymbol']) and (quote == product['quoteSymbol'])):
+            if ((target_price is None) or (target_price.buy_price is None) or (target_price.sell_price is None)):
+                base_allocation = Wad(0)
```
Comment (MikeHathaway) [inline market_maker_keeper/leverjfutures_market_maker_keeper_v2.py:217]: Likely want to be using `Wad.from_number()` here as `Wad.from_number(0) != Wad(0)`

File: market_maker_keeper/leverjfutures_market_maker_keeper_v2.py (hunk 5 above)
Comment (mitakash) [reply market_maker_keeper/leverjfutures_market_maker_keeper_v2.py:196]: We initially tried to trade multiple pairs using same account and then scaled back to 1 account for 1 pair.

File: market_maker_keeper/leverjfutures_market_maker_keeper_v2.py (hunk 1 above)
Comment (mitakash) [reply market_maker_keeper/leverjfutures_market_maker_keeper_v2.py:172]: this is a good idea, will default this to 1 when we intially take in arguments.

File: market_maker_keeper/leverjfutures_market_maker_keeper_v2.py (hunk 2 above)
Comment (mitakash) [reply market_maker_keeper/leverjfutures_market_maker_keeper_v2.py:272]: yup, will add typings to this

File: market_maker_keeper/leverjfutures_market_maker_keeper_v2.py (hunk 3 above)
Comment (mitakash) [reply market_maker_keeper/leverjfutures_market_maker_keeper_v2.py:42]: Not sure about this, need further discussion with you about how to implement this and if it's a good idea

File: market_maker_keeper/leverjfutures_market_maker_keeper_v2.py (hunk 4 above)
Comment (mitakash) [reply market_maker_keeper/leverjfutures_market_maker_keeper_v2.py:184]: This piece of code is trying to calculate max possible tokens we can take on. It's fairly static and is based on the max DAI amount in the wallet

File: market_maker_keeper/leverjfutures_market_maker_keeper_v2.py (hunk 6 above)
Comment (mitakash) [reply market_maker_keeper/leverjfutures_market_maker_keeper_v2.py:210]: We need a general idea of what the current price is to do a bunch of calculations. basically what's really in the wallet is just DAI, we basically use half that DAI to "estimate" how much fake Base token we have to be able to place sell orders

File: market_maker_keeper/leverjfutures_market_maker_keeper_v2.py (hunk 7 above)
Comment (mitakash) [reply market_maker_keeper/leverjfutures_market_maker_keeper_v2.py:217]: This is a good idea, we could change the code to Wad.from_number(0)

Comment (mitakash) [reply market_maker_keeper/leverjfutures_market_maker_keeper_v2.py:217]: so leaving as is Wad(0) is equivalent to Wad.from_number(0)

**Review by MikeHathaway — APPROVED** [review]
LGTM

**Pull request merged**